File: server-side/src/api/product/product.controller.ts
import { Body, Controller, Delete, Get, Param, Post, Put } from '@nestjs/common';
import { ProductService } from './product.service';
import { CreateProductDto } from './dto/create-product.dto';
import { UpdateProductDto } from './dto/update-product.dto';
import { Product } from './entities/product.entity';

@Controller('product')
export class ProductController {
  constructor(private readonly productService: ProductService) {}

  @Get('/all')
  async getAllProducts(): Promise<Product[]> {
    return this.productService.getAllProducts();
  }

  @Get('/single/:id')
  async getProductById(@Param('id') id: number): Promise<Product | null> {
    return this.productService.getProductById(id);
  }

  @Post('/create')
  async createProduct(@Body() body: CreateProductDto): Promise<Product> {
    return this.productService.createProduct(body);
  }

  @Put('/update/:id')
  async updateProduct(
    @Param('id') id: number,
    @Body() body: UpdateProductDto,
  ): Promise<Product | null> {
    return this.productService.updateProduct(id, body);
  }

  @Delete('/delete/:id')
  async deleteProduct(@Param('id') id: number): Promise<{ success: boolean }> {
    const success = await this.productService.deleteProduct(id);
    return { success };
  }
}

File: backend-server/src/main/java/DM/com/backend_server/Controller/ProductController.java
package DM.com.backend_server.Controller;

import DM.com.backend_server.Entity.Product;
import DM.com.backend_server.dto.CreateProductDto;
import DM.com.backend_server.dto.UpdateProductDto;
import DM.com.backend_server.service.ProductService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

// @CrossOrigin(origins = "http://localhost:3001")
@RestController
@RequestMapping("/api/v1/product")
public class ProductController {

    private final ProductService productService;

    @Autowired
    public ProductController(ProductService productService) {
        this.productService = productService;
    }

    @GetMapping("/all")
    public ResponseEntity<List<Product>> getAllProducts() {
        List<Product> products = productService.getAllProducts();
        return ResponseEntity.ok(products);
    }

    @GetMapping("/single/{id}")
    public ResponseEntity<Product> getProductById(@PathVariable Long id) {
        Product product = productService.getProductById(id);
        return ResponseEntity.ok(product);
    }

    @PostMapping("create")
    public ResponseEntity<Product> createProduct(@RequestBody CreateProductDto data) {
        Product product = productService.createProduct(data);
        return ResponseEntity.status(HttpStatus.CREATED).body(product);
    }

    @PutMapping("/update/{id}")
    public ResponseEntity<Product> updateProduct(@PathVariable Long id, @RequestBody UpdateProductDto data) {
        Product updatedProduct = productService.updateProduct(id, data);
        return ResponseEntity.ok(updatedProduct);
    }

    @DeleteMapping("/delete/{id}")
    public ResponseEntity<Void> deleteProduct(@PathVariable Long id) {
        boolean deleted = productService.deleteProduct(id);
        if (deleted) {
            return ResponseEntity.noContent().build();
        }
        return ResponseEntity.notFound().build(); // Fallback, though exception would handle it
    }
}
